### PR TITLE
Enable accumulation checkers to accumulate facts on type variables (and use this new facility in the RLC)

### PR DESCRIPTION
**Pull request by kelloggm**
Inspired by the similar facility in the KeyFor checker, which is used by the Nullness Checker.

Note that the RLC doesn't actually use the `getAccumulatedValues(Tree)` method that I added. I wanted to include such a method, though, so that anyone else building an accumulation checker in the future has a simple access point to use this feature. I'm open to removing it, if a reviewer thinks that we shouldn't add it (since it isn't strictly required).

### CODE REVIEW

**Comment by msridhar** [comment]
@kelloggm if you'd like my review in addition to @smillst's, it'd be helpful if you could point out where the main logic changes are here.  I think there are a lot of changes that are just renames or the like.

**Comment by kelloggm** [comment]
@msridhar I'd like you to review the changes to `MustCallConsistencyAnalyzer`, especially the ones around line 2033 that actually use the new functionality introduced by this PR, and the associated test cases (`tests/resourceleak/TypevarSimple.java` and `tests/resourceleak/Issue4815.java`). Sorry that so many boiler-plate changes were required.

If you're willing to dig into it, I'd also appreciate your thoughts on the design of the feature (i.e., `AccumulationValue.java`), but I think Suzanne is better equipped to give me feedback there.

File: framework/src/main/java/org/checkerframework/common/accumulation/AccumulationValue.java
Comment (smillst) [inline framework/src/main/java/org/checkerframework/common/accumulation/AccumulationValue.java:20]: This isn't true. `Accumulator(a) T` is a supertype of `Accumulator(a,b) T` but `T` is not a supertype of `Accumulator(a) T` unless both bounds of `T` are supertypes of of `Accumulator(a)`.

**Review by smillst — CHANGES_REQUESTED** [review]
Code looks good, but there needs to be more tests.  Please see the comment I added.

File: framework/src/main/java/org/checkerframework/common/accumulation/AccumulationValue.java
Comment (smillst) [inline framework/src/main/java/org/checkerframework/common/accumulation/AccumulationValue.java:109]: This is returning greatest lower bound, but it should be returning:
if `this <: other`: return this;
if `other <: this`: return other;
otherwise: return backup;

Comment (kelloggm) [reply framework/src/main/java/org/checkerframework/common/accumulation/AccumulationValue.java:109]: This code is based on very similar code in `KeyForValue`, which also seems (to me) to be computing a GLB. Is that code also incorrect?

Debugging this code on the failing test cases, I don't see that it's doing anything wrong. The interesting case is when `mostSpecific` is called on the following values:
```
this: CFAV{@CalledMethods("a"), T, [a] }
other: CFAV{, T, null }
```
In that case, the result is what I'd expect: `CFAV{, T, [a] }`

Would you expect something different here? Maybe I'm misunderstanding the specification of `mostSpecific`?

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:2031]: Do we know why this variable is called `aliasCFValue`, in particular what `CF` stands for?  If not, maybe we should rename?

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:2036]: As I understand it `accumulatedValues` can only be non-null if we have a type variable or wildcard type.  While this is documented very clearly elsewhere it may be helpful to have a short comment here also:

```suggestion
        if (accumulatedValues != null) { // type variable or wildcard type
```

**Review by msridhar — DISMISSED** [review]
I just have minor comments on the parts of the code I was asked to review, which don't need to block landing of this PR.  @smillst should approve before this lands though

File: framework/src/main/java/org/checkerframework/common/accumulation/AccumulationValue.java
Comment (smillst) [reply framework/src/main/java/org/checkerframework/common/accumulation/AccumulationValue.java]: I was wrong.  The problem was that values were added to `this` and `other` rather than created a new mostSpecific value.  I've made that change.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java
Comment (kelloggm) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:2031]: I think "CF" there is based on the old type (`CFValue`). I've changed the name of the variable to `cmValue`, which parallels `cmAnno` on the next line (both are for `alias`, which is another local, but I think that's 1) implied from context, and 2) not a great name itself.

**Pull request merged**
